articleview.cc

```diff
@@ -2669,6 +2661,27 @@ void ArticleView::highlightFTSResults()
   ftsSearchIsOpened = true;
 }
 
+void ArticleView::highlightAllFtsOccurences( QWebPage::FindFlags flags )
+{
+  flags |= QWebPage::HighlightAllOccurrences;
+
+  // Usually allMatches contains mostly duplicates. Thus searching for each element of
+  // allMatches to highlight them takes a long time => collect unique elements into a
+  // set and search for them instead.
+  // Don't use QList::toSet() or QSet's range constructor because they reserve space
+  // for QList::size() elements, whereas the final QSet size is likely 1 or 2.
+  QSet< QString > uniqueMatches;
```

Comment:
I personally prefer the SmallSet you implemented in mediawiki.cc, as it has inspired me a lot. Is there a big difference in performance if change QSet to SmallSet and merge the two for loops into one?
https://github.com/goldendict/goldendict/blob/df192bf555c04c0cd36eeb774f40988f9bbdedfd/mediawiki.cc#L250

Reply:
Merging the two loops into one is probably a bad idea as it would mix different RAM access patterns, which could negatively impact CPU cache utilization.

Using `SmallSet` risks ruining performance in corner cases here. A FTS regular expression may match lots of different words, in which case `SmallSet` would run in quadratic time.

Reply:
Yes, results produced by a regular expression may be huge even if duplicates removed. I really did not take into account the case of using regular expression.